File: src/main/java/com/customagent/service/AnalysisService.java
package com.customagent.service;

import com.customagent.exception.AgentNotFoundException;
import com.customagent.exception.AnalysisException;
import com.customagent.model.Agent;
import com.customagent.model.AgentStatus;
import com.customagent.model.AnalysisReport;
import com.customagent.model.ReportStatus;
import com.customagent.model.dto.AnalysisRequest;
import com.customagent.repository.AgentRepository;
import com.customagent.repository.AnalysisReportRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class AnalysisService {

    private final AnalysisReportRepository reportRepository;
    private final AgentRepository          agentRepository;

    // ─── TRIGGER ─────────────────────────────────────────────────────────────

    /**
     * Starts a new analysis job for the given repository using the specified agent.
     *
     * @throws AgentNotFoundException if the agent does not exist
     * @throws AnalysisException      if the agent is not ACTIVE
     */
    public AnalysisReport triggerAnalysis(AnalysisRequest request) {
        Agent agent = agentRepository.findById(request.getAgentId())
                .orElseThrow(() -> new AgentNotFoundException(
                        "Agent not found with id: " + request.getAgentId()));

        if (agent.getStatus() != AgentStatus.ACTIVE) {
            throw new AnalysisException(
                    "Cannot trigger analysis with inactive agent: " + agent.getId());
        }

        AnalysisReport report = AnalysisReport.builder()
                .repositoryUrl(request.getRepositoryUrl())
                .agentId(agent.getId())
                .status(ReportStatus.PENDING)
                .build();

        AnalysisReport saved = reportRepository.save(report);
        log.info("Triggered analysis reportId={} agentId={} url={}",
                saved.getId(), saved.getAgentId(), saved.getRepositoryUrl());
        return saved;
    }

    // ─── READ ─────────────────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public AnalysisReport getReportById(Long id) {
        return reportRepository.findById(id)
                .orElseThrow(() -> new AnalysisException(
                        "Analysis report not found with id: " + id));
    }

    @Transactional(readOnly = true)
    public List<AnalysisReport> getReportsByAgentId(Long agentId) {
        return reportRepository.findByAgentId(agentId);
    }

    @Transactional(readOnly = true)
    public List<AnalysisReport> getReportsByStatus(ReportStatus status) {
        return reportRepository.findByStatus(status);
    }

    // ─── STATE TRANSITIONS ────────────────────────────────────────────────────

    /**
     * Marks a report as COMPLETED and stores the result payload.
     *
     * @throws AnalysisException if the report is already in a terminal state
     */
    public AnalysisReport completeAnalysis(Long reportId, String result) {
        AnalysisReport report = getReportById(reportId);

        if (report.getStatus() == ReportStatus.COMPLETED) {
            throw new AnalysisException("Report " + reportId + " is already completed");
        }
        if (report.getStatus() == ReportStatus.FAILED) {
            throw new AnalysisException(
                    "Cannot complete a failed report: " + reportId);
        }

        report.setStatus(ReportStatus.COMPLETED);
        report.setResult(result);
        report.setCompletedAt(LocalDateTime.now());
        return reportRepository.save(report);
    }

    /**
     * Marks a report as FAILED and stores the error message.
     *
     * @throws AnalysisException if the report is already in a terminal state
     */
    public AnalysisReport failAnalysis(Long reportId, String errorMessage) {
        AnalysisReport report = getReportById(reportId);

        if (report.getStatus() == ReportStatus.COMPLETED
                || report.getStatus() == ReportStatus.FAILED) {
            throw new AnalysisException(
                    "Cannot fail report " + reportId
                            + " in terminal status: " + report.getStatus());
        }

        report.setStatus(ReportStatus.FAILED);
        report.setResult(errorMessage);
        report.setCompletedAt(LocalDateTime.now());
        return reportRepository.save(report);
    }
}

File: src/services/analysisService.js
'use strict';

/**
 * Analysis Service — manages code-analysis job lifecycle.
 * Mirrors the Java AnalysisService business logic.
 *
 * State machine:
 *   PENDING → COMPLETED
 *   PENDING → FAILED
 *   (terminal states cannot transition)
 */

const ReportStatus = Object.freeze({
  PENDING: 'PENDING',
  COMPLETED: 'COMPLETED',
  FAILED: 'FAILED',
});

let _nextId = 1;

/**
 * Creates an AnalysisService bound to the provided repositories.
 *
 * @param {{ findById, findByAgentId, findByStatus, save }} reportRepository
 * @param {{ findById }} agentRepository
 * @returns {object} service instance
 */
function createAnalysisService(reportRepository, agentRepository) {
  // ── TRIGGER ─────────────────────────────────────────────────────────────

  /**
   * Starts a new analysis job.
   *
   * @param {{ repositoryUrl: string, agentId: number }} request
   * @returns {object} Saved report with PENDING status
   * @throws {Error} If agent not found or agent is not ACTIVE
   */
  function triggerAnalysis(request) {
    if (!request || !request.repositoryUrl || request.repositoryUrl.trim() === '') {
      throw new Error('Repository URL is required');
    }
    if (request.agentId === null || request.agentId === undefined) {
      throw new Error('Agent ID is required');
    }

    const agent = agentRepository.findById(request.agentId);
    if (!agent) {
      throw new Error(`Agent not found with id: ${request.agentId}`);
    }
    if (agent.status !== 'ACTIVE') {
      throw new Error(`Cannot trigger analysis with inactive agent: ${agent.id}`);
    }

    const report = {
      id: _nextId++,
      repositoryUrl: request.repositoryUrl.trim(),
      agentId: agent.id,
      status: ReportStatus.PENDING,
      result: null,
      createdAt: new Date().toISOString(),
      completedAt: null,
    };

    return reportRepository.save(report);
  }

  // ── READ ─────────────────────────────────────────────────────────────────

  /**
   * @param {number} id
   * @returns {object} Report
   * @throws {Error} If not found
   */
  function getReportById(id) {
    const report = reportRepository.findById(id);
    if (!report) {
      throw new Error(`Analysis report not found with id: ${id}`);
    }
    return report;
  }

  /**
   * @param {number} agentId
   * @returns {object[]}
   */
  function getReportsByAgentId(agentId) {
    return reportRepository.findByAgentId(agentId);
  }

  /**
   * @param {string} status - One of ReportStatus values
   * @returns {object[]}
   */
  function getReportsByStatus(status) {
    return reportRepository.findByStatus(status);
  }

  // ── STATE TRANSITIONS ────────────────────────────────────────────────────

  /**
   * Marks a PENDING report as COMPLETED.
   *
   * @param {number} reportId
   * @param {string} result  JSON-serialised result payload
   * @returns {object} Updated report
   * @throws {Error} If report is in a terminal state
   */
  function completeAnalysis(reportId, result) {
    const report = getReportById(reportId);

    if (report.status === ReportStatus.COMPLETED) {
      throw new Error(`Report ${reportId} is already completed`);
    }
    if (report.status === ReportStatus.FAILED) {
      throw new Error(`Cannot complete a failed report: ${reportId}`);
    }

    report.status = ReportStatus.COMPLETED;
    report.result = result;
    report.completedAt = new Date().toISOString();
    return reportRepository.save(report);
  }

  /**
   * Marks a PENDING report as FAILED.
   *
   * @param {number} reportId
   * @param {string} errorMessage
   * @returns {object} Updated report
   * @throws {Error} If report is already in a terminal state
   */
  function failAnalysis(reportId, errorMessage) {
    const report = getReportById(reportId);

    if (report.status === ReportStatus.COMPLETED || report.status === ReportStatus.FAILED) {
      throw new Error(
        `Cannot fail report ${reportId} in terminal status: ${report.status}`
      );
    }

    report.status = ReportStatus.FAILED;
    report.result = errorMessage;
    report.completedAt = new Date().toISOString();
    return reportRepository.save(report);
  }

  return {
    triggerAnalysis,
    getReportById,
    getReportsByAgentId,
    getReportsByStatus,
    completeAnalysis,
    failAnalysis,
  };
}

/** Reset auto-increment id (useful for testing). */
function resetIdCounter(start = 1) {
  _nextId = start;
}

module.exports = { createAnalysisService, ReportStatus, resetIdCounter };
